src/hazelcore/extractions/ocaml/OCamlExtraction_Exp.re

```
@@ -0,0 +1,270 @@
+      }
+    | _ => failwith("Exp: Cons")
+    };
+  | Inj(t, side, de) =>
+    // t is the type of another side, for example int | bool, side = L, then t = bool
+    // FIXME: do we really need to restrict t can't be Hole? It means we should explicitly declear the both side of sum type, i.e., let x : Int | Bool = ...
+    // injection is the introductory of sum type in hazel, as designed in Typ, we pack them as `Left and `Right constructor
+    let ocaml_exp = extract(~ctx, ~de);
+    switch (side) {
+    | L => ("(`Left " ++ fst(ocaml_exp) ++ ")", Sum(snd(ocaml_exp), t))
+    | R => ("(`Right " ++ fst(ocaml_exp) ++ ")", Sum(t, snd(ocaml_exp)))
+    };
+  | Pair(d1, d2) =>
+    let ocaml_d1 = extract(~ctx, ~de=d1);
+    let ocaml_d2 = extract(~ctx, ~de=d2);
+    (
+      "(" ++ fst(ocaml_d1) ++ ", " ++ fst(ocaml_d2) ++ ")",
+      Prod([snd(ocaml_d1), snd(ocaml_d2)]),
+    );
+  | Triv => ("()", Hole) //We have no unit Htype...
+  | ConsistentCase(cases) =>
+    switch (cases) {
+    // the int seems useless
+    | Case(de, rules, _) =>
+      let ocaml_de = extract(~ctx, ~de);
+      let ocaml_rules = rules_extract(~ctx, ~rules, ~pat_t=snd(ocaml_de));
+      //FIXME: Do we really need to indicate the type?
+      let str =
+        "((match ("
+        ++ fst(ocaml_de)
+        ++ ") with \n"
+        ++ fst(ocaml_rules)
+        ++ ") : "
+        ++ OCamlExtraction_Typ.extract(~t=snd(ocaml_rules))
+        ++ ")";
+      (str, snd(ocaml_rules));
+    }
+  // inconsistbranches is a case have inconsistent types
+  | InconsistentBranches(_, _, _, _) =>
+    failwith("Exp: Ocaml don't allow inconsistent branches")
+  | Cast(_, _, _) => failwith("Exp: cast not allowed")
+  | FailedCast(_, _, _) => failwith("Exp: cast not allowed")
+  | InvalidOperation(_, _) => failwith("Exp: Invalid Operation")
+  }
+and rules_extract =
+    (~ctx: Contexts.t, ~rules: list(DHExp.rule), ~pat_t: HTyp.t): t =>
+  switch (rules) {
+  | [] => ("", Hole) //no rules, won't happen actually
```

should raise an exception if it is an impossible case